---
title: "Vietnamese children's microbiome in diarrhoea episode"
author: "Son-Nam H. Le"
date: "11/29/2021"
output: pdf_document
---
Initialize
```{r}
library(phyloseq)
library(stringr)
library(dplyr)
library(phytools)
library(microViz)
library(ggplot2)
library("cluster")
library(class)
library(randomForestSRC)
library(ggRandomForests)
library(fpc)
library(PERMANOVA)
library(qiime2R)
library(microbiome)
library(microbiomeutilities)
library(ggpubr)
library(ranacapa)
library(vegan)
library(RColorBrewer)
library(htmltools)
library(ggtree)
set.seed(100)
```

Load file
```{r}
load(file = "D:/Study/Oucru/SonNam_project/diarrhea_dataset/diarrhea_dataset/phyloseq_select_1904-OTUs_fin.RData")
phylotree_all <- read.newick(file = "D:/Study/Oucru/SonNam_project/diarrhea_dataset/diarrhea_dataset/select_filter_1904-OTUs_pasta_trimal_gt5%gap.fasta.reroot.newick")

```

Edit 
```{r}


filter.ps <- merge_phyloseq(filter.ps, phylotree_all)
  
metadata <- data.frame(sample_data(filter.ps))

allgroup.taxglom <- tax_glom(filter.ps, taxrank = "Genus", NArm = FALSE)
dt <- transform_sample_counts(allgroup.taxglom, function(OTU) OTU/sum(OTU))
```
Weighted Unifrac distance (old)
```{r}
wunifrac.dist <- phyloseq::distance(dt, method="wunifrac")
ord.wunifrac <- ordinate(dt, method="PCoA", distance=wunifrac.dist)
plot_scree(ord.wunifrac) + xlim(as.character(seq(1,50))) + coord_cartesian(xlim=c(0,50)) + ggtitle("Weighted Unifrac distance based principal coordinate scree plot") + xlab("Principal coordinate") + ylab("Proportion of variance")
dt@sam_data[["day"]] <- as.factor(dt@sam_data[["day"]])
plot_ordination(dt, ord.wunifrac, color="day") + stat_ellipse() + ggtitle("Weighted Unifrac")  + theme_classic()

### Using Gap statistic, and clustering directly on distance matrix 
gsy = clusGap(t(otu_table(dt)), pam, K.max=10, B=100)
plot_clusgap(gsy) + scale_x_continuous(breaks=c(seq(1,10,1)))
```

Calculate by another method (Paul Igor Costea) (old)
```{r}
pred <- function (distance, Gmin = 2, Gmax = 10, M = 50, 
classification = "centroid", cutoff = 0.8, nnk = 1, ...) 
{
require(cluster)
require(class)
xdata <- as.matrix(distance)
n <- nrow(xdata)
nf <- c(floor(n/2), n - floor(n/2))
indvec <- clcenters <- clusterings <- jclusterings <- classifications <- list()
prederr <- list()
dist <- as.matrix(distance)
for (k in Gmin:Gmax) {
    prederr[[k]] <- numeric(0)
    for (l in 1:M) {
        nperm <- sample(n, n)
        indvec[[l]] <- list()
        indvec[[l]][[1]] <- nperm[1:nf[1]]
        indvec[[l]][[2]] <- nperm[(nf[1] + 1):n]
        for (i in 1:2) {
            clusterings[[i]] <- as.vector(pam(as.dist(dist[indvec[[l]][[i]],indvec[[l]][[i]]]), k, diss=TRUE))
            jclusterings[[i]] <- rep(-1, n)
            jclusterings[[i]][indvec[[l]][[i]]] <- clusterings[[i]]$clustering
    centroids <- clusterings[[i]]$medoids
            j <- 3 - i
            classifications[[j]] <- classifdist(as.dist(dist), jclusterings[[i]], 
              method = classification, centroids = centroids, 
              nnk = nnk)[indvec[[l]][[j]]]
        }
        ps <- matrix(0, nrow = 2, ncol = k)
        for (i in 1:2) {
            for (kk in 1:k) {
              nik <- sum(clusterings[[i]]$clustering == kk)
              if (nik > 1) {
                for (j1 in (1:(nf[i] - 1))[clusterings[[i]]$clustering[1:(nf[i] - 
                  1)] == kk]) {
                  for (j2 in (j1 + 1):nf[i]) if (clusterings[[i]]$clustering[j2] == 
                    kk) 
                    ps[i, kk] <- ps[i, kk] + (classifications[[i]][j1] == 
                      classifications[[i]][j2])
                }
                ps[i, kk] <- 2 * ps[i, kk]/(nik * (nik - 
                  1))
              }
            }
        }
        prederr[[k]][l] <- mean(c(min(ps[1, ]), min(ps[2, 
            ])))
    }
}
mean.pred <- numeric(0)
if (Gmin > 1) 
    mean.pred <- c(1)
if (Gmin > 2) 
    mean.pred <- c(mean.pred, rep(NA, Gmin - 2))
for (k in Gmin:Gmax) mean.pred <- c(mean.pred, mean(prederr[[k]]))
optimalk <- max(which(mean.pred > cutoff))
out <- list(predcorr = prederr, mean.pred = mean.pred, optimalk = optimalk, 
    cutoff = cutoff, method = clusterings[[1]]$clustermethod, 
    Gmax = Gmax, M = M)
class(out) <- "predstr"
out
}
#Validating the best number of clustering
PS.y <- pred(distance = wunifrac.dist, 2, 10, 50, clustermethod=pamkCBI, classification="centroid", cutoff=0.60, count=TRUE) # 4 clusters
asw.y <- pamk(wunifrac.dist, krange=2:10, criterion="asw", diss=TRUE,ns=10, critout = TRUE) # 4 clusters
BS.y <- nselectboot(wunifrac.dist, B = 50, clustermethod = pamkCBI, classification = "averagedist", krange=2:10) # 4 clusters

plot(pam(wunifrac.dist,4, diss=TRUE))

```
Assessing the ability to assign samples to correct group by assessing the sample silhouette width. 
```{r}
K <- 4
sample_data(dt)$CST <- as.factor(pam(x=wunifrac.dist, K, cluster.only =TRUE, diss=TRUE))
clust <- as.array(setNames(as.numeric(sample_data(dt)$CST), rownames(sample_data(dt))))
clust.al <- clust
silwidths <- silhouette(clust.al, wunifrac.dist)
rownames(silwidths) <- names(clust.al)
print(silwidths)

### For sample with si < 0, change the cluster membership to the nearest neighbor
clust.al["AHH20074"] <- 3
clust.al["AHH20095"] <- 2
clust.al["AHH20154"] <- 3
clust.al["AHH20164"] <- 3
clust.al["AHH20643"] <- 2
clust.al["AHH20645"] <- 2
clust.al["AHH20783"] <- 4
clust.al["AHH20827"] <- 4
clust.al["PHH2379"] <- 2
clust.al["PHH2812"] <- 4
clust.al["PHH2829"] <- 4

### comparing between the two assignment, clust and clust.al

plot(silhouette(clust, wunifrac.dist))
clust.al <- as.array(clust.al)
plot(silhouette(clust.al, wunifrac.dist)) 
print(silhouette(clust.al, wunifrac.dist))

### Define group membership 
sample_data(dt)$CST <- as.factor(clust.al)
sample_data(filter.ps)$CST <- as.factor(clust.al)
```

WUni Diversity by group
```{r}
physeq_rel <- microbiome::transform(filter.ps, "compositional")
physeq.ord.wuni <- ordinate(physeq_rel, "PCoA", "unifrac", weighted=T)

#Weighted Unifrac

b.div.wuni2 <- plot_ordination(physeq_rel, physeq.ord.wuni, type= "samples", color= "CST") + geom_point(size=1)
b.div.wuni2 <- b.div.wuni2 + stat_ellipse() + ggtitle("Weighted Unifrac")  + theme_classic()
print(b.div.wuni2)

plot_ordination(physeq_rel, physeq.ord.wuni, color="CST", axes = c(2,3))+ stat_ellipse()
```

Random forest
```{r}
class.df <- as.data.frame(t(otu_table(dt)))
class.df$CST <- as.factor(clust.al)

class.rf <- rfsrc(CST~., data = class.df,ntree=10000) # random forest classification based on all  genus level OTUs. 
class.vimp <- vimp.rfsrc(class.rf)
print(class.rf) ### overall error rate is 11.54%
class.rf.OOBerror <- gg_error(class.rf)
plot.rfsrc(class.rf) ### plotting individual error rate for each CST
plot(gg_rfsrc(class.rf)) ### visualizing the correct prediction rate for each CST

### Selecting only OTUs that significantly contribute to the classification process
A <- as.data.frame(class.vimp$importance >0.2) 
otus.class <- rownames(A[!(A$all=="FALSE" & A$`1` =="FALSE" & A$`2`=="FALSE" & A$`3`=="FALSE" & A$`4`=="FALSE"),])
all.class <- prune_taxa(otus.class,allgroup.taxglom )
all.taxglom.abund <- transform_sample_counts(all.class, function(OTU) OTU/sum(OTU))

class.df <- as.data.frame(t(otu_table(all.taxglom.abund)))
class.df$CST <- as.factor(clust.al)

class.rf <- rfsrc(CST~., data = class.df, ntree=10000) ### random forest classification based on only 28 most important OTUs
class.vimp <- vimp.rfsrc(class.rf)
print(class.rf) ## overall error rate is about 12.98%, not so much differrent from using all data
class.rf.OOBerror <- gg_error(class.rf)
plot.rfsrc(class.rf) ### stabilized error rate, with individual CST's error rate close to each other, Figure S1
plot(gg_rfsrc(class.rf)) 
plot(gg_vimp(vimp.rfsrc(class.rf))) #CST 1 does not have the most presenter
```

Assessing, read phyloseq file
```{r}

library(metagMisc)

#read file
ntaxa(filter.ps)
nsamples(filter.ps)
sample_names(filter.ps)[1:5]  
rank_names(filter.ps)  
sample_variables(filter.ps)  
otu_table(filter.ps)[1:5, 1:5]  
tax_table(filter.ps)[1:5, 1:4]
```

Plot composition
```{r}
#filter all patient which are recorded at 3 time points
data.04en <- subset_samples(filter.ps, study_ID == "04EN")
df.patient <- data.04en@sam_data %>% group_by(patient_ID) %>% summarise_at(vars(c(day)), funs(sum(.,narm=TRUE)))
df.patient3 <- filter(df.patient, day == 23)
patientID.3d <- pull(df.patient3, patient_ID)
patient3d <- subset_samples(data.04en, patient_ID %in% patientID.3d)


#change the type of "day", turn "NA" to healthy
filter.ps@sam_data[["day"]] <- as.factor(filter.ps@sam_data[["day"]])
patient3d@sam_data[["day"]] <- as.factor(patient3d@sam_data[["day"]])

filter.ps@sam_data[["day"]] = factor(filter.ps@sam_data[["day"]], levels=c(levels(filter.ps@sam_data[["day"]]), "healthy"))
filter.ps@sam_data[["day"]][is.na(filter.ps@sam_data[["day"]])] = "healthy"

#change type of some data
patient3d@sam_data[["diarrtimes_day"]] <- as.integer(patient3d@sam_data[["diarrtimes_day"]])
patient3d@sam_data[["diarrhea_stool"]] <- as.factor(patient3d@sam_data[["diarrhea_stool"]])



#pick up top family patient 3d:
patient3d.fam <- microbiome::aggregate_top_taxa(patient3d, "Family", top = 7)
patient3d.fam.rel <- transform(patient3d.fam, "compositional")
patient3d.rel.abund <- plot_composition(patient3d.fam.rel, sample.sort = "diarrtimes_day") + facet_wrap(~sample_data(patient3d.fam.rel)$day)
patient3d.rel.abund$data$day <- sample_data(patient3d.fam.rel)$day
patient3d.rel.abund$data$dday <- sample_data(patient3d.fam.rel)$diarrtimes_day
patient3d.rel.abund$data$dstool <- sample_data(patient3d.fam.rel)$diarrhea_stool
data.com <- patient3d.rel.abund$data

data.com <- arrange(data.com, dday)


  #heat maps
  p.heat <- ggplot(data.com, aes(x = Sample, y = Tax)) + geom_tile(aes(fill = Abundance))
  

  # Change color
  p.heat <- p.heat + theme_bw() 

  # Make bacterial names italics
  p.heat <- p.heat + theme(axis.text.y = element_text(colour = 'black', 
                                                    size = 10, 
                                                    face = 'italic'))
  #by day
  p.heat <- p.heat +facet_grid(~dstool)
  
  #clean x axis
  p.heat <- p.heat + theme(axis.title.x=element_blank(),
                     axis.text.x=element_blank(),
                     axis.ticks.x=element_blank()) 
  
  print(p.heat)
  

  #dot
  p.dot <- ggplot(data.com, aes(x = Sample, y = dday)) + geom_point(aes(color = Tax, size = Abundance, shape = dstool ), alpha = 0.8, stroke = 1) + yscale("log2") + labs(title = "Diarrhoea stool and Abundance families", x = "Samples by the division of hospitalisation days", y = "Diarrhoea times per day", shape = "Diarrhoea stool", color = "Family")
  # Change color
  p.dot <- p.dot + theme_classic() 

  # Make bacterial names italics
  p.dot <- p.dot + theme(axis.text.y = element_text(colour = 'black', 
                                                    size = 10, 
                                                    face = 'italic'))
  #by day
  p.dot <- p.dot + facet_wrap(~day)
  
  #clean x axis
  p.dot <- p.dot + theme(axis.text.x=element_blank(),
                     axis.ticks.x=element_blank()) 
  
  print(p.dot)
```


```{r}
#rarefy the phyloseq object to even depth prior various analysis
physeq_rarefy <- rarefy_even_depth(filter.ps, rngseed=1, sample.size=0.9*min(sample_sums(filter.ps)), replace=F)

#create the new function "plot_taxa_prevalence2"
plot_taxa_prevalence2 <- function (x, level, detection = 0) {
  abundance <- NULL
  prevalence <- NULL
  #x <- check_phyloseq(x, fill_na_taxa = TRUE)
    tax.abun <- apply(otu_table(x), 1, mean)
    tax.prev <- prevalence(x, detection = detection)
    Phylum <- as.vector(data.frame(tax_table(x))$Phylum)
    Phylum <- as.vector(Phylum)
    Phylum <- as.factor(Phylum)
    xdf <- data.frame(abundance = log(tax.abun, base=10), prevalence = tax.prev, 
                      Phylum)
    plot.phylum <- ggplot(xdf, aes(x = abundance, y = prevalence, 
                                   color = Phylum)) + geom_point(shape = 16, alpha = 0.9) + 
      xlab("Average count abundance (log scale)") + 
      ylab("Taxa prevalence") + theme_bw() + theme(plot.background = element_blank(), 
                                                   panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
                                                   panel.background = element_blank(), axis.text.x = element_text(angle = 90, 
                                                                                                                  vjust = 0.5, size = 6)) + geom_vline(xintercept = 0, 
                                                                                                                                                       linetype = "dashed", color = "grey") + 
      facet_wrap(~Phylum) + theme(strip.background = element_rect(fill = "white"))
    return(plot.phylum)
}
#plot phylum prevalence
plot_taxa_prevalence2(filter.ps, "Phylum")

  #composition plot for patient (3 time points)
  physeq_fam.3d <- microbiome::aggregate_rare(patient3d.atrt.y, level = "Family", detection = 30/100, prevalence = 50/100)
  physeq.fam.rel.f.3d <- microbiome::transform(physeq_fam.f, "compositional")

  physeq.fam.rel.f.3d <- patient3d %>%
    aggregate_rare(level = "Family", detection = 30/100, prevalence = 50/100) %>%
    microbiome::transform(transform = "compositional")

  plot_composition(physeq.fam.rel.f.3d,sample.sort = "day", x.label = "patient_ID", group_by = "day") + theme(legend.position = "bottom") +   scale_fill_brewer("Family", palette = "Paired") + theme_bw() + theme(axis.text.x = element_text(angle = 90)) + ggtitle("Relative abundance") + theme(legend.title = element_text(size = 15))

#composition plot for patient
physeq_fam <- microbiome::aggregate_rare(filter.ps, level = "Family", detection = 30/100, prevalence = 50/100)
physeq.fam.rel <- microbiome::transform(physeq_fam, "compositional")

physeq.fam.rel <- filter.ps %>%
  aggregate_rare(level = "Family", detection = 30/100, prevalence = 50/100) %>%
  microbiome::transform(transform = "compositional")

plot_composition(physeq.fam.rel,sample.sort = "day", x.label = "day") + theme(legend.position = "bottom") + scale_fill_brewer("Family", palette = "Paired") + theme_bw() + theme(axis.text.x = element_text(angle = 90)) + ggtitle("Relative abundance") + theme(legend.title = element_text(size = 15))

```

Most abundant taxa:

```{r}
physeq_df <- microbiomeutilities::phy_to_ldf(physeq_fam, 
                                         transform.counts = "compositional")

#an additonal column Sam_rep with sample names is created for reference purpose
colnames(physeq_df)

#box plot at family level

ggboxplot(physeq_df, "day", "Abundance", 
             facet.by = "Family", color = "day",
             palette = "Set1") + rremove("x.text")

#plot specific taxa
physeq.f <- format_to_besthit(filter.ps)

top_taxa(physeq.f, 4)

# If more than two variables
comps <- make_pairs(sample_data(physeq.f)$day)
print(comps)
# plot some specific taxa
select.taxa <- c("Otu00001:Bifidobacterium longum", "Otu00003:Escherichia coli", "Otu00597:Streptococcus salivarius")

mycols <- c("coral", "steelblue2", "slategray2", "olivedrab")

p <- plot_listed_taxa(physeq.f, select.taxa, 
                 group= "day",
                 group.order = c(1, 7, 14, "Healthy"),
                 group.colors = mycols,
                 add.violin = F,
                 dot.opacity = 0.25,
                 box.opacity = 0.25,
                 panel.arrange= "grid") + ylab("Relative abundance") + scale_y_continuous(labels = scales::percent)

p <- p + stat_compare_means(
  comparisons = comps,
  label = "p.format",
  tip.length = 0.05,
  method = "wilcox.test") 

p + scale_y_continuous(labels = scales::percent)

#top genera

#physeq.genus <- aggregate_taxa(filter.ps, "Family") #subcript out of bounds
physeq.genus <- tax_glom(filter.ps, taxrank="Genus")
top_four <- top_taxa(physeq.genus, 4)
print(top_four)

top_genera <- plot_listed_taxa(physeq.genus, top_four, 
                 group= "day",
                 group.order = c(1, 7, 14, "NA"),
                 group.colors = mycols,
                 add.violin = F,
                 dot.opacity = 0.25,
                 box.opacity = 0.25,
                 panel.arrange= "wrap")

top_genera + stat_compare_means(
                   comparisons = comps,
                   label = "p.format",
                   label.y = 50,
                   tip.length = 0.05,
                   method = "wilcox.test")

```
Alpha diversities
```{r}
plot_richness(filter.ps, measures="Shannon")


a.div <- plot_richness(physeq_rarefy, x="day", measures=c("Shannon", "simpson", "Observed"), color = "diarrhea_stool") + geom_boxplot() + theme_bw() + aes(group = day)

# adding statistical support
plot.a.div <- a.div + stat_compare_means(
  comparisons = comps,
  label = "p.signif",
  tip.length = 0.05,
  symnum.args = list(
    cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1), 
    symbols = c("****", "***", "**", "*", "ns")
  ),
  method = "wilcox.test") 

print(plot.a.div)
```

Beta diversity metrices
```{r}

#non-phylogenetic beta diversity metrics (Bray - Curtis)

physeq.ord <- ordinate(physeq_rel, "PCoA", "bray")
b.div.bray <- plot_ordination(physeq_rel, physeq.ord, type= "samples", color= "day") + geom_point(size=1)
b.div.bray <- b.div.bray + stat_ellipse() + ggtitle("Bray Curtis")  + theme_classic() 
print(b.div.bray)

#Weighted Unifrac
physeq.ord.wuni <- ordinate(physeq_rel, "PCoA", "unifrac", weighted=T)
physeq_rel@sam_data[["day"]] <- as.factor(physeq_rel@sam_data[["day"]])
b.div.wuni <- plot_ordination(physeq_rel, physeq.ord.wuni, type= "samples", color= "day") + geom_point(size=1)
b.div.wuni <- b.div.wuni + stat_ellipse() + ggtitle("Weighted Unifrac")  + theme_classic()
print(b.div.wuni)

#PERMANOVA
  otu <- abundances(physeq_rel)
  meta <- meta(physeq_rel)
  dist.bray <- vegdist(t(otu), "bray")

  #statistics - Beta diversity (xem lai function adonis, totu thanh distance matrix) 
    permanova <- adonis(t(dist.bray) ~ meta$day, data = meta, permutations=99, method = "bray")

  #P-value
    print(as.data.frame(permanova$aov.tab)["day", "Pr(>F)"])
  #Pair - wise stats
    anova(betadisper(dist.bray, meta$day))
    permutest(betadisper(dist.bray, meta$day), pairwise = TRUE)
#homogeneity condition
    #Pair - wise stats
    anova(betadisper(dist, meta$day))
    permutest(betadisper(dist, meta$day), pairwise = TRUE)

```
Core Microbiota
```{r}
physeq.1 <- subset_samples(filter.ps, day == "1")

# convert to relative abundance  
physeq.1.rel <- microbiome::transform(physeq.1, "compositional")

physeq.1.rel2 <- prune_taxa(taxa_sums(physeq.1.rel) > 0, physeq.1.rel)

#core in each phase
core.taxa.standard.1 <- core_members(physeq.1.rel2, detection = 0.001, prevalence = 50/100)
print(core.taxa.standard.1)

# Extract the taxonomy table
taxonomy_core.1 <- as.data.frame(tax_table(physeq.1.rel2))

# Subset this taxonomy table to include only core OTUs
core_taxa_id.1 <- subset(taxonomy_core.1, rownames(taxonomy_core.1) %in% core.taxa.standard.1)

DT::datatable(core_taxa_id.1)
```

Interaction network
```{r}
library(SpiecEasi)

#filter before running the SpiecEasi)

otu_table(physeq.f)

top20 <- sort(taxa_sums(physeq.f), TRUE)[1:20]

top20.f <- prune_taxa(names(top20), physeq.f)

#SpiecEasi
otu.net <- t(otu_table(top20.f)@.Data)
tax <- as.data.frame(tax_table(top20.f)@.Data)

net <- spiec.easi(otu.net, method='mb', lambda.min.ratio=1e-2, nlambda=20, pulsar.params=list(rep.num=50))

graph.net <- adj2igraph((getRefit(net)))

```

